### PR TITLE
chore: pin docker images by hash

### PR DESCRIPTION
**Pull request by crenshaw-dev**
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,10 +1,10 @@
-ARG BASE_IMAGE=docker.io/library/ubuntu:22.04
+ARG BASE_IMAGE=docker.io/library/ubuntu:22.04@sha256:c985bc3f77946b8e92c9a3648c6f31751a7dd972e06604785e47303f4ad47c4c
```
Comment (34fathombelow) [inline Dockerfile:1]:
```suggestion
ARG BASE_IMAGE=docker.io/library/ubuntu:22.04@sha256:9a0bdde4188b896a372804be2384015e90e3f84906b750c1a53539b585fbbe7f
```

File: Dockerfile [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
 ####################################################################################################
 # Builder image
 # Initial stage which pulls prepares build dependencies and CLI tooling we need for our final image
 # Also used as the image in CI jobs so needs all dependencies
 ####################################################################################################
-FROM docker.io/library/golang:1.20 AS builder
+FROM docker.io/library/golang:1.20@sha256:2edf6aab2d57644f3fe7407132a0d1770846867465a39c2083770cf62734b05d AS builder
```
Comment (34fathombelow) [inline Dockerfile:7]:
```suggestion
FROM docker.io/library/golang:1.19.6@sha256:7ce31d15a3a4dbf20446cccffa4020d3a2974ad2287d96123f55caf22c7adb71 AS builder
```

File: Dockerfile [hunk 3]
```diff
@@ -101,7 +101,7 @@ RUN HOST_ARCH=$TARGETARCH NODE_ENV='production' NODE_ONLINE_ENV='online' NODE_OP
 ####################################################################################################
 # Argo CD Build stage which performs the actual build of Argo CD binaries
 ####################################################################################################
-FROM --platform=$BUILDPLATFORM docker.io/library/golang:1.20 AS argocd-build
+FROM --platform=$BUILDPLATFORM docker.io/library/golang:1.20@sha256:2edf6aab2d57644f3fe7407132a0d1770846867465a39c2083770cf62734b05d AS argocd-build
```
Comment (34fathombelow) [inline Dockerfile:104]:
```suggestion
FROM --platform=$BUILDPLATFORM docker.io/library/golang:1.19.6@sha256:7ce31d15a3a4dbf20446cccffa4020d3a2974ad2287d96123f55caf22c7adb71 AS argocd-build
```

File: test/container/Dockerfile [hunk 4]
```diff
@@ -1,19 +1,19 @@
...
 # There are libraries we will want to copy from here in the final stage of the
 # build, but the COPY directive does not have a way to determine system
 # architecture, so we create a symlink here to facilitate copying.
 RUN ln -s /usr/lib/$(uname -m)-linux-gnu /usr/lib/linux-gnu 
 
-FROM docker.io/library/node:12.18.4-buster as node
+FROM docker.io/library/node:12.18.4-buster@sha256:7f4090a3fea631e703f4afe5cc5504143970d669a2c13b4e5edf3f830c9b4c96 as node
 
-FROM docker.io/library/golang:1.20 as golang
+FROM docker.io/library/golang:1.20@sha256:2edf6aab2d57644f3fe7407132a0d1770846867465a39c2083770cf62734b05d as golang
```
Comment (34fathombelow) [inline test/container/Dockerfile:10]:
```suggestion
FROM docker.io/library/golang:1.19.6@sha256:7ce31d15a3a4dbf20446cccffa4020d3a2974ad2287d96123f55caf22c7adb71 as golang
```

File: test/container/Dockerfile [hunk 5]
```diff
@@ -1,19 +1,19 @@
...
 
-FROM docker.io/library/registry:2.8 as registry
+FROM docker.io/library/registry:2.8@sha256:3f71055ad7c41728e381190fee5c4cf9b8f7725839dcf5c0fe3e5e20dc5db1fa as registry
 
-FROM docker.io/bitnami/kubectl:1.24.8 as kubectl
+FROM docker.io/bitnami/kubectl:1.24.8@sha256:9c1bd1fe7bf4a0c19b4b38e596f355c5172805e14e6a3f71433472e3456e86d8 as kubectl
```
Comment (34fathombelow) [inline test/container/Dockerfile:14]:
```suggestion
FROM docker.io/bitnami/docker.io/bitnami/kubectl:1.26@sha256:625467eb8c3a3d60232923404941c32e787eb9003e644d0fa8258b0efa7f6a7f as kubectl
```

File: test/container/Dockerfile [hunk 6]
```diff
@@ -1,19 +1,19 @@
...
 
-FROM ubuntu:22.04
+FROM ubuntu@sha256:9a0bdde4188b896a372804be2384015e90e3f84906b750c1a53539b585fbbe7f
```
Comment (34fathombelow) [inline test/container/Dockerfile:16]:
```suggestion
FROM docker.io/library/ubuntu:22.04@sha256:9a0bdde4188b896a372804be2384015e90e3f84906b750c1a53539b585fbbe7f
```

File: test/remote/Dockerfile [hunk 7]
```diff
@@ -1,6 +1,6 @@
 ARG BASE_IMAGE=docker.io/library/ubuntu:22.04
 
-FROM golang:1.20 AS go
+FROM golang:1.20@sha256:2edf6aab2d57644f3fe7407132a0d1770846867465a39c2083770cf62734b05d AS go
```
Comment (34fathombelow) [inline test/remote/Dockerfile:3]:
```suggestion
FROM docker.io/library/golang:1.19.6@sha256:7ce31d15a3a4dbf20446cccffa4020d3a2974ad2287d96123f55caf22c7adb71 AS go
```

File: ui-test/Dockerfile [hunk 8]
```diff
@@ -1,4 +1,4 @@
-FROM node:12.18.4 AS node
+FROM node:12.18.4@sha256:8cfe7e8dc60095a4f9d25a3f0f208503559fa033a15e2ddd87dee85bec101a2e AS node
```
Comment (34fathombelow) [inline ui-test/Dockerfile:1]:
```suggestion
FROM docker.io/library/node:12.18.4@sha256:8cfe7e8dc60095a4f9d25a3f0f208503559fa033a15e2ddd87dee85bec101a2e AS node
```

File: .github/dependabot.yml [hunk 9]
```diff
@@ -16,3 +16,8 @@ updates:
     directory: "/ui/"
     schedule:
       interval: "daily"
+
+  - package-ecosystem: "docker"
+    directory: "/"
+    schedule:
+      interval: "daily"
```
Comment (34fathombelow) [inline .github/dependabot.yml:23]: We need to add `package-ecosystem: "docker"` for each directory we want dependabot to scan.  Unfortunately we cannot pass multiple directories as a list.

**Review by 34fathombelow — CHANGES_REQUESTED** [review]
All other changes look good.

File: test/container/Dockerfile [hunk 10]
```diff
@@ -1,19 +1,19 @@
...
+FROM docker.io/library/node:12.18.4-buster@sha256:7f4090a3fea631e703f4afe5cc5504143970d669a2c13b4e5edf3f830c9b4c96 as node
 
-FROM docker.io/library/golang:1.20 as golang
+FROM docker.io/library/golang:1.19.6@sha256:7ce31d15a3a4dbf20446cccffa4020d3a2974ad2287d96123f55caf22c7adb71 as golang
 
-FROM docker.io/library/registry:2.8 as registry
+FROM docker.io/library/registry:2.8@sha256:3f71055ad7c41728e381190fee5c4cf9b8f7725839dcf5c0fe3e5e20dc5db1fa as registry
 
-FROM docker.io/bitnami/kubectl:1.24.8 as kubectl
+FROM docker.io/bitnami/docker.io/bitnami/kubectl:1.26@sha256:625467eb8c3a3d60232923404941c32e787eb9003e644d0fa8258b0efa7f6a7f as kubectl
```
Comment (crenshaw-dev) [inline test/container/Dockerfile:14]:
```suggestion
FROM docker.io/bitnami/kubectl:1.26@sha256:625467eb8c3a3d60232923404941c32e787eb9003e644d0fa8258b0efa7f6a7f as kubectl
```

**Review by 34fathombelow — APPROVED** [review]
Lgtm

Just note we need to be cautious when dependabot recommends updates for Golang and node versions.  We should limit these to patch or minor versions.

**Review by jannfis — APPROVED** [review]
LGTM!

**Pull request merged**